File: src/989_add_to_array_form_of_integer/add_to_array_form_of_integer.py
from typing import List


class Solution:
    def addToArrayForm(self, num: List[int], k: int) -> List[int]:
        carry = 0
        i = len(num) - 1
        res = []

        while k or carry or i >= 0:
            sum = carry + (k % 10)
            if i >= 0:
                sum += num[i]

            res.append(sum % 10)
            k //= 10
            carry = sum // 10
            i -= 1

        res.reverse()

        return res


# Test scaffolding
if __name__ == "__main__":
    tests = [
        (([1, 2, 0, 0], 34), [1, 2, 3, 4]),
        (([2, 7, 4], 181), [4, 5, 5]),
        (([2, 1, 5], 806), [1, 0, 2, 1]),
        (([0], 23), [2, 3]),
        (([9, 9, 9, 9, 9], 1), [1, 0, 0, 0, 0, 0]),
    ]
    sol = Solution()
    for (num, k), expected in tests:
        actual = sol.addToArrayForm(num, k)
        print(f"{actual} | {expected}")

File: src/989_add_to_array_form_of_integer/add_to_array_form_of_integer.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
   public:
    vector<int> addToArrayForm(vector<int>& num, int k) {
        int i = num.size() - 1;
        int carry = 0;
        vector<int> res;
        while (k > 0 || carry > 0 || i >= 0) {
            int sum = carry + (k % 10);
            if (i >= 0) sum += num[i];
            res.push_back(sum % 10);
            k /= 10;
            carry = sum / 10;
            i--;
        }
        reverse(res.begin(), res.end());
        return res;
    }
};

int main() {
    vector<pair<vector<int>, int>> tests = {
        {{1, 2, 0, 0}, 34}, {{2, 7, 4}, 181}, {{2, 1, 5}, 806}, {{0}, 23}, {{9, 9, 9, 9, 9}, 1}};
    vector<vector<int>> expected = {
        {1, 2, 3, 4}, {4, 5, 5}, {1, 0, 2, 1}, {2, 3}, {1, 0, 0, 0, 0, 0}};
    Solution sol;
    for (int i = 0; i < tests.size(); ++i) {
        auto actual = sol.addToArrayForm(tests[i].first, tests[i].second);
        cout << "[";
        for (size_t j = 0; j < actual.size(); ++j) {
            cout << actual[j];
            if (j + 1 < actual.size()) cout << ",";
        }
        cout << "] | [";
        for (size_t j = 0; j < expected[i].size(); ++j) {
            cout << expected[i][j];
            if (j + 1 < expected[i].size()) cout << ",";
        }
        cout << "]" << endl;
    }
    return 0;
}
